src/test_utils.rs

```diff
@@ -0,0 +1,203 @@
+            ToPeer::NoConnection | ToPeer::Initiated { .. } => {
+                panic!(
+                    "Peer {} is in invalid state {:?} to be communicated to",
+                    peer_addr, conn.to_peer
+                );
+            }
+        }
+    })
+}
+
+/// Mock `QuicP2p`, used for testing malicious client/node scenarios.
+pub(crate) struct TestPeer {
+    event_tx: Sender<Event>,
+    el: EventLoop,
+}
+
+impl TestPeer {
+    /// Create a test peer
+    pub fn new(event_tx: Sender<Event>) -> Self {
+        let el = EventLoop::spawn();
+        Self { event_tx, el }
+    }
+
+    /// Must be called only once. There can only be one context per `TestPeer` instance.
+    pub fn activate(&mut self, our_type: OurType) -> R<()> {
+        let tx = self.event_tx.clone();
+
+        let ((key, cert), our_complete_cert) = {
+            let our_complete_cert: SerialisableCertificate = Default::default();
+            (
+                our_complete_cert.obtain_priv_key_and_cert(),
+                our_complete_cert,
+            )
+        };
+
+        let (err_tx, err_rx) = mpsc::channel::<R<()>>();
+        self.el.post(move || {
+            let our_cfg = unwrap!(peer_config::new_our_cfg(
+                peer_config::DEFAULT_IDLE_TIMEOUT_MSEC,
+                peer_config::DEFAULT_KEEP_ALIVE_INTERVAL_MSEC,
+                cert,
+                key
+            ));
+
+            let mut ep_builder = quinn::Endpoint::builder();
+            ep_builder.listen(our_cfg);
+            let (dr, ep, incoming_connections) = {
+                match UdpSocket::bind(&(Ipv4Addr::LOCALHOST, 0)) {
+                    Ok(udp) => unwrap!(ep_builder.with_socket(udp)),
+                    Err(e) => {
+                        panic!("Could not bind a random port! Error: {:?}- {}", e, e);
+                    }
+                }
+            };
+
+            let mut bootstrap_cache =
+                unwrap!(BootstrapCache::new(Default::default(), Default::default()));
+
+            // Do not use the bootstrap cache.
+            mem::replace(bootstrap_cache.peers_mut(), Default::default());
+
+            let ctx = Context::new(
+                tx,
+                our_complete_cert,
+                crate::DEFAULT_MAX_ALLOWED_MSG_SIZE,
+                peer_config::DEFAULT_IDLE_TIMEOUT_MSEC,
+                peer_config::DEFAULT_KEEP_ALIVE_INTERVAL_MSEC,
+                our_type,
+                bootstrap_cache,
+                ep,
+            );
+
+            initialise_ctx(ctx);
+
+            current_thread::spawn(dr.map_err(|e| warn!("Error in quinn Driver: {:?}", e)));
+
+            if our_type != OurType::Client {
+                listener::listen(incoming_connections);
+            }
+
+            let _ = err_tx.send(Ok(()));
+        });
+
+        (err_rx.recv()?)?;
+
+        Ok(())
+    }
+
+    /// Send an arbitrary message.
+    pub fn send(&mut self, peer: Peer, msg: WireMsg) {
```

Comment:
If everything else is the same and only this is different can we now just _add_ this function for `QuicP2p` as an imp in tests ?

like in `test_utils`:
```
impl QuicP2p {
    fn send_wire_msg(...) { .. }
}
```

Reply:
Ah that's a nice trick I totally forgot about. Yeah, it should work and we don't need an explicit `activate` fn taking `OurType` either because it could be always provided in the config. :+1: